### PR TITLE
pdi: init at 1.7.1

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Description of changes

Add support for [pdi](https://pdi.dev/master/index.html), loose coupling of simulation codes with data handling libraries.

- **python312Packages.zpp, zpp: init at 1.1.0**
- **pdi: init at 1.7.1**

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 344977`

---
### `x86_64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>python311Packages.zpp</li>
    <li>python311Packages.zpp.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>pdi</li>
    <li>zpp (python312Packages.zpp)</li>
    <li>zpp.dist (python312Packages.zpp.dist)</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>python311Packages.zpp</li>
    <li>python311Packages.zpp.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>pdi</li>
    <li>zpp (python312Packages.zpp)</li>
    <li>zpp.dist (python312Packages.zpp.dist)</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>pdi</li>
  </ul>
</details>
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.zpp</li>
    <li>python311Packages.zpp.dist</li>
    <li>zpp (python312Packages.zpp)</li>
    <li>zpp.dist (python312Packages.zpp.dist)</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>pdi</li>
  </ul>
</details>
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.zpp</li>
    <li>python311Packages.zpp.dist</li>
    <li>zpp (python312Packages.zpp)</li>
    <li>zpp.dist (python312Packages.zpp.dist)</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 344977`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 5 packages built:</summary>
  <ul>
    <li>pdi</li>
    <li>python311Packages.zpp</li>
    <li>python311Packages.zpp.dist</li>
    <li>zpp (python312Packages.zpp)</li>
    <li>zpp.dist (python312Packages.zpp.dist)</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>python311Packages.zpp</li>
    <li>python311Packages.zpp.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>zpp (python312Packages.zpp)</li>
    <li>zpp.dist (python312Packages.zpp.dist)</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.zpp</li>
    <li>python311Packages.zpp.dist</li>
    <li>zpp (python312Packages.zpp)</li>
    <li>zpp.dist (python312Packages.zpp.dist)</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.zpp</li>
    <li>python311Packages.zpp.dist</li>
    <li>zpp (python312Packages.zpp)</li>
    <li>zpp.dist (python312Packages.zpp.dist)</li>
  </ul>
</details>

File: pkgs/by-name/zp/zpp/package.nix [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{ python3Packages }:
```
Comment (SuperSandro2000) [inline pkgs/by-name/zp/zpp/package.nix:1]: Why not place this in all-packages.nix?

Comment (GaetanLepage) [reply pkgs/by-name/zp/zpp/package.nix:1]: Isn't that the new recommended way ?
If not I can for sure put it back in `all-packages.nix` .

Comment (GaetanLepage) [reply pkgs/by-name/zp/zpp/package.nix:1]: @SuperSandro2000, hexa advised me favor the `by-name` approach.

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 344977`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 5 packages built:</summary>
  <ul>
    <li>pdi</li>
    <li>python311Packages.zpp</li>
    <li>python311Packages.zpp.dist</li>
    <li>zpp (python312Packages.zpp)</li>
    <li>zpp.dist (python312Packages.zpp.dist)</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.zpp</li>
    <li>python311Packages.zpp.dist</li>
    <li>zpp (python312Packages.zpp)</li>
    <li>zpp.dist (python312Packages.zpp.dist)</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.zpp</li>
    <li>python311Packages.zpp.dist</li>
    <li>zpp (python312Packages.zpp)</li>
    <li>zpp.dist (python312Packages.zpp.dist)</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>python311Packages.zpp</li>
    <li>python311Packages.zpp.dist</li>
    <li>zpp (python312Packages.zpp)</li>
    <li>zpp.dist (python312Packages.zpp.dist)</li>
  </ul>
</details>

File: pkgs/by-name/zp/zpp/package.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/by-name/zp/zpp/package.nix:1]: > Isn't that the new recommended way ?

I thought that is for new packages like pdi, not for "aliases", short hands or multi version packages where I think this exactly falls into it.

Comment (GaetanLepage) [reply pkgs/by-name/zp/zpp/package.nix:1]: Yes, I thought too. This is why I wanted to ask hexa, and he said "just use the by-name approach, dealing with all-packages.nix is too annoying".

Comment (GaetanLepage) [reply pkgs/by-name/zp/zpp/package.nix:1]: Ok, I will go with the current implementation then.

While I was looking to create a PR to document this choice, I realized that this is already the recommended process: https://nixos.org/manual/nixpkgs/unstable/#topythonapplication-function

Comment (GaetanLepage) [reply pkgs/by-name/zp/zpp/package.nix:1]: Thank you once again for your careful review ! I appreciate :)

Comment (SuperSandro2000) [reply pkgs/by-name/zp/zpp/package.nix:1]: Glad that I am just hearing now from all the changes in the python language document.

**Pull request merged**